### PR TITLE
Add table aws_dynamodb_cluster (DAX).closes #315

### LINKED ISSUE #315: Add table aws_dynamodb_cluster (DAX)
**References**
Use case here is ```[DynamoDB.3] This control checks whether a DAX cluster is encrypted at rest```

https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html#dynamodb-3-remediation

### PR DESCRIPTION
**Pull request by Priyanka585464**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_dax_cluster []

PRETEST: tests/aws_dax_cluster

TEST: tests/aws_dax_cluster
Running terraform
data.aws_partition.current: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_region.alternate: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_iam_role.my_role: Creating...
aws_vpc.my_vpc: Creating...
aws_iam_role.my_role: Still creating... [10s elapsed]
aws_vpc.my_vpc: Still creating... [10s elapsed]
aws_vpc.my_vpc: Still creating... [20s elapsed]
aws_iam_role.my_role: Still creating... [20s elapsed]
aws_iam_role.my_role: Still creating... [30s elapsed]
aws_vpc.my_vpc: Still creating... [30s elapsed]
aws_vpc.my_vpc: Still creating... [40s elapsed]
aws_iam_role.my_role: Still creating... [40s elapsed]
aws_iam_role.my_role: Creation complete after 47s [id=turbottest37938]
aws_vpc.my_vpc: Still creating... [50s elapsed]
aws_vpc.my_vpc: Still creating... [1m0s elapsed]
aws_vpc.my_vpc: Still creating... [1m10s elapsed]
aws_vpc.my_vpc: Still creating... [1m20s elapsed]
aws_vpc.my_vpc: Still creating... [1m30s elapsed]
aws_vpc.my_vpc: Still creating... [1m40s elapsed]
aws_vpc.my_vpc: Still creating... [1m50s elapsed]
aws_vpc.my_vpc: Still creating... [2m0s elapsed]
aws_vpc.my_vpc: Creation complete after 2m3s [id=vpc-0da9f28c756a982ab]
aws_subnet.my_subnet1: Creating...
aws_subnet.my_subnet2: Creating...
aws_subnet.my_subnet2: Still creating... [10s elapsed]
aws_subnet.my_subnet1: Still creating... [10s elapsed]
aws_subnet.my_subnet2: Still creating... [20s elapsed]
aws_subnet.my_subnet1: Still creating... [20s elapsed]
aws_subnet.my_subnet2: Still creating... [30s elapsed]
aws_subnet.my_subnet1: Still creating... [30s elapsed]
aws_subnet.my_subnet1: Creation complete after 34s [id=subnet-0ace51e4b8c40d790]
aws_subnet.my_subnet2: Creation complete after 34s [id=subnet-0e88c1d37ac50bfc1]
aws_dax_subnet_group.my_subnet_group: Creating...
aws_dax_subnet_group.my_subnet_group: Still creating... [10s elapsed]
aws_dax_subnet_group.my_subnet_group: Still creating... [20s elapsed]
aws_dax_subnet_group.my_subnet_group: Creation complete after 23s [id=turbottest37938]
aws_dax_cluster.named_test_resource: Creating...
aws_dax_cluster.named_test_resource: Still creating... [10s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [20s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [30s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [40s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [50s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [1m0s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [1m10s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [1m20s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [1m30s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [1m40s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [1m50s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [2m0s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [2m10s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [2m20s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [2m30s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [2m40s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [2m50s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [3m0s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [3m10s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [3m20s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [3m30s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [3m40s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [3m50s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [4m0s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [4m10s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [4m20s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [4m30s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [4m40s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [4m50s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [5m0s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [5m10s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [5m20s elapsed]
aws_dax_cluster.named_test_resource: Still creating... [5m30s elapsed]
aws_dax_cluster.named_test_resource: Creation complete after 5m33s [id=turbottest37938]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 6 added, 0 changed, 0 destroyed.

Outputs:

account_id = 352685002396
aws_partition = aws
aws_region = us-east-1
resource_aka = arn:aws:dax:us-east-1:352685002396:cache/turbottest37938
resource_id = turbottest37938
resource_name = turbottest37938

Running SQL query: test-get-query.sql
[
  {
    "arn": "arn:aws:dax:us-east-1:352685002396:cache/turbottest37938",
    "cluster_name": "turbottest37938",
    "description": "A test cluster",
    "node_type": "dax.r4.large",
    "tags": {
      "name": "turbottest37938"
    }
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:dax:us-east-1:352685002396:cache/turbottest37938"
    ],
    "cluster_name": "turbottest37938",
    "tags": {
      "name": "turbottest37938"
    },
    "title": "turbottest37938"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "akas": [
      "arn:aws:dax:us-east-1:352685002396:cache/turbottest37938"
    ],
    "cluster_name": "turbottest37938",
    "title": "turbottest37938"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

POSTTEST: tests/aws_dax_cluster

TEARDOWN: tests/aws_dax_cluster

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

```sql
select
  cluster_name,
  description,
  active_nodes,
  iam_role_arn,
  status,
  region
from
  aws_dax_cluster;
```
```
+----------------------+-------------+--------------+-----------------------------------------------------+-----------+-----------+
| cluster_name         | description | active_nodes | iam_role_arn                                        | status    | region    |
+----------------------+-------------+--------------+-----------------------------------------------------+-----------+-----------+
| turbot-test-20200125 | <null>      | 1            | arn:aws:iam::986325076436:role/turbot-test-20200125 | available | us-east-1 |
+----------------------+-------------+--------------+-----------------------------------------------------+-----------+-----------+
```

### List clusters that does not enforce server-side encryption (SSE)

```sql
select
  cluster_name,
  description,
  sse_description ->> 'Status' as sse_status
from
  aws_dax_cluster
where
  sse_description ->> 'Status' = 'DISABLED';
```
```
+----------------------+-------------+------------+
| cluster_name         | description | sse_status |
+----------------------+-------------+------------+
| turbot-test-20200125 | <null>      | DISABLED   |
+----------------------+-------------+------------+
```

### List clusters provisioned with undesired (for example, dax.r4.large and dax.t2.medium are desired) node types

```sql
select
  cluster_name,
  node_type,
  count(*) as count
from
  aws_dax_cluster
where
  node_type not in ('cache.m5.large', 'cache.m4.4xlarge')
group by
  cluster_name, node_type;
```
```
+----------------------+--------------+-------+
| cluster_name         | node_type    | count |
+----------------------+--------------+-------+
| turbot-test-20200125 | dax.r4.large | 1     |
+----------------------+--------------+-------+
```

### Get the network details for each cluster

```sql
select
  cluster_name,
  subnet_group,
  sg ->> 'SecurityGroupIdentifier' as sg_id,
  n ->> 'AvailabilityZone' as az_name,
  cluster_discovery_endpoint ->> 'Address' as cluster_discovery_endpoint_address,
  cluster_discovery_endpoint ->> 'Port' as cluster_discovery_endpoint_port
from
  aws_dax_cluster,
  jsonb_array_elements(security_groups) as sg,
  jsonb_array_elements(nodes) as n
```
```
+----------------------+----------------------+----------------------+------------+---------------------------------------------------------------------+---------------------------------+
| cluster_name         | subnet_group         | sg_id                | az_name    | cluster_discovery_endpoint_address                                  | cluster_discovery_endpoint_port |
+----------------------+----------------------+----------------------+------------+---------------------------------------------------------------------+---------------------------------+
| turbot-test-20200125 | turbot-test-20200125 | sg-00ef710282a054a53 | us-east-1e | turbot-test-20200125.z6keng.clustercfg.dax.use1.cache.amazonaws.com | 8111                            |
+----------------------+----------------------+----------------------+------------+---------------------------------------------------------------------+---------------------------------+
```

### CODE REVIEW

File: aws-test/tests/aws_dax_cluster/test-get-expected.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+	{
+		"arn": "{{ output.resource_aka.value }}",
+		"cluster_name": "{{resourceName}}",
+		"description": "A test cluster",
+		"node_type": "dax.r4.large",
+		"tags": {
+			"name": "{{resourceName}}"
+		}
+	}
+]
```
Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/test-get-expected.json:11]: Fix indentation in json file

File: aws-test/tests/aws_dax_cluster/test-hydrate-expected.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[
+	{
+		"akas": ["{{ output.resource_aka.value }}"],
+		"cluster_name": "{{ output.resource_name.value }}",
+		"tags": {
+			"name": "{{resourceName}}"
+		},
+		"title": "{{ output.resource_name.value }}"
+	}
+]
```
Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/test-hydrate-expected.json:10]: Fix indentation in json file

File: aws-test/tests/aws_dax_cluster/test-get-expected.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[
+	{
+		"arn": "{{ output.resource_aka.value }}",
+		"cluster_name": "{{resourceName}}",
```
Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/test-get-expected.json:4]:
```suggestion
		"cluster_name": "{{ resourceName }}",
```

File: aws-test/tests/aws_dax_cluster/test-get-expected.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+		"description": "A test cluster",
+		"node_type": "dax.r4.large",
+		"tags": {
+			"name": "{{resourceName}}"
```
Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/test-get-expected.json:8]:
```suggestion
			"name": "{{ resourceName }}"
```

File: aws-test/tests/aws_dax_cluster/test-hydrate-expected.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[
+	{
+		"akas": ["{{ output.resource_aka.value }}"],
+		"cluster_name": "{{ output.resource_name.value }}",
+		"tags": {
+			"name": "{{resourceName}}"
```
Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/test-hydrate-expected.json:6]:
```suggestion
			"name": "{{ resourceName }}"
```

File: aws-test/tests/aws_dax_cluster/test-list-expected.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+[
+	{
+		"akas": ["{{ output.resource_aka.value }}"],
+		"cluster_name": "{{ output.resource_name.value }}",
+		"title": "{{resourceName}}"
```
Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/test-list-expected.json:5]:
```suggestion
		"title": "{{ resourceName }}"
```

File: aws-test/tests/aws_dax_cluster/test-list-expected.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+	}
+]
```
Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/test-list-expected.json:7]: Fix indentation in json file

File: aws-test/tests/aws_dax_cluster/variables.tf [hunk 8]
```diff
@@ -0,0 +1,120 @@
+variable "resource_name" {
```
Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/variables.tf:1]:
```suggestion

variable "resource_name" {
```

Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/variables.tf:1]:
```suggestion


```

Comment (anisadas) [inline aws-test/tests/aws_dax_cluster/variables.tf:1]:
```suggestion

variable "resource_name" {
```

File: aws/table_aws_dax_cluster.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+	pagesLeft := true
+	params := &dax.DescribeClustersInput{}
+
+	for pagesLeft {
+		result, err := svc.DescribeClusters(params)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, clusters := range result.Clusters {
```
Comment (anisadas) [inline aws/table_aws_dax_cluster.go:175]:
```suggestion
		for _, cluster := range result.Clusters {
```

File: aws/table_aws_dax_cluster.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+			d.StreamListItem(ctx, clusters)
```
Comment (anisadas) [inline aws/table_aws_dax_cluster.go:176]:
```suggestion
			d.StreamListItem(ctx, cluster)
```

File: aws/table_aws_dax_cluster.go [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+	clusterdata, err := svc.ListTags(params)
+	if err != nil {
+		return nil, err
+	}
+
+	return clusterdata, nil
+}
+
+// TRANSFORM FUNCTION
```
Comment (anisadas) [inline aws/table_aws_dax_cluster.go:252]:
```suggestion
//// TRANSFORM FUNCTION
```

File: aws/table_aws_dax_cluster.go [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+
+func daxClusterTurbotData(_ context.Context, d *transform.TransformData) (interface{}, error) {
+	data := d.HydrateItem.(*dax.ListTagsOutput)
+	if data.Tags == nil {
+		return nil, nil
+	}
+
+	// Get the resource tags
```
Comment (anisadas) [inline aws/table_aws_dax_cluster.go:260]:
```suggestion
	// Mapping the resource tags inside turbotTags
```

**Comment by anisadas** [comment]
@Priyanka585464 / @Paulami30 Make suggested changes

**Review by Paulami30 — APPROVED** [review]

**Review by Paulami30 — COMMENTED** [review]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**